### PR TITLE
[NPU]Optimization of `forward_npu` for `UnquantizedFusedMoEMethod`

### PR DESCRIPTION
**Pull request by Vladimir221**
## Motivation

Using of `torch_npu.npu_moe_init_routing_v2` instead of sequence of operations: `torch_npu.npu_moe_init_routing` + `torch_npu.npu_moe_compute_expert_tokens` improves performance of non quantized MoE models

## Modifications

Implementation of `forward_npu` method of `UnquantizedFusedMoEMethod` class has been changed, `torch_npu.npu_moe_init_routing` + `torch_npu.npu_moe_compute_expert_tokens` sequence has been substituted by `torch_npu.npu_moe_init_routing` also weights transposition has been moved to `process_weights_after_loading` method of the mentioned class

## Accuracy Tests
**Ascend 910B, tp-size=4, concurrency=64, GSM8k**

_Before:_

```
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [03:32<00:00,  6.21it/s]
Accuracy: 0.911
Invalid: 0.001
Latency: 212.462 s
Output throughput: 740.286 token/s
```

_After:_
```
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [03:08<00:00,  7.01it/s]
Accuracy: 0.906
Invalid: 0.000
Latency: 188.337 s
Output throughput: 809.818 token/s
```

## Benchmarking and Profiling

_Before:_

Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 64        
Successful requests:                     64        
Benchmark duration (s):                  32.98     
Total input tokens:                      8192      
Total input text tokens:                 8192      
Total input vision tokens:               0         
Total generated tokens:                  65536     
Total generated tokens (retokenized):    65535     
Request throughput (req/s):              1.94      
Input token throughput (tok/s):          248.40    
Output token throughput (tok/s):         1987.17   
Total token throughput (tok/s):          2235.57   
Concurrency:                             63.93     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   32945.42  
Median E2E Latency (ms):                 32946.66  
---------------Time to First Token----------------
Mean TTFT (ms):                          493.27    
Median TTFT (ms):                        495.69    
P99 TTFT (ms):                           507.06    
---------------Inter-Token Latency----------------
Mean ITL (ms):                           31.72     
Median ITL (ms):                         31.69     
P95 ITL (ms):                            33.06     
P99 ITL (ms):                            34.16     
Max ITL (ms):                            208.15    

_After:_

Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 64        
Successful requests:                     64        
Benchmark duration (s):                  28.59     
Total input tokens:                      8192      
Total input text tokens:                 8192      
Total input vision tokens:               0         
Total generated tokens:                  65536     
Total generated tokens (retokenized):    65531     
Request throughput (req/s):              2.24      
Input token throughput (tok/s):          286.56    
Output token throughput (tok/s):         2292.48   
Total token throughput (tok/s):          2579.05   
Concurrency:                             63.93     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   28554.61  
Median E2E Latency (ms):                 28554.86  
---------------Time to First Token----------------
Mean TTFT (ms):                          484.57    
Median TTFT (ms):                        487.50    
P99 TTFT (ms):                           498.66    
---------------Inter-Token Latency----------------
Mean ITL (ms):                           27.44     
Median ITL (ms):                         27.34     
P95 ITL (ms):                            28.96     
P99 ITL (ms):                            29.83     
Max ITL (ms):                            218.21    

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Review by ssshinigami — APPROVED** [review]
LGTM

**Review by ping1jing2 — APPROVED** [review]

File: python/sglang/srt/layers/quantization/unquant.py [hunk 1]
```diff
@@ -231,6 +237,18 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
         if _is_cpu and _is_cpu_amx_available:
             _amx_process_weight_after_loading(layer, ["w13_weight", "w2_weight"])
 
+        if _is_npu:
+            for weight_name in ["w13_weight", "w2_weight"]:
+                weight = getattr(layer, weight_name)
+                origin_data = weight.data
+                weight.data = torch_npu.npu_format_cast(
+                    origin_data, NPU_FORMAT_FRACTAL_NZ
+                )
```
Comment (iforgetmyname) [inline python/sglang/srt/layers/quantization/unquant.py]: we have a utility function to do this operation here at `srt.hardware_backend.npu.utils.npu_format_cast`

File: python/sglang/srt/layers/quantization/unquant.py [hunk 2]
```diff
@@ -542,6 +554,7 @@ def forward_npu(
             scales=topk_weights,
             expanded_src_to_dst_row=expanded_row_idx,
             export_for_source_row=topk_ids,
+            drop_pad_mode=2,
```
Comment (iforgetmyname) [inline python/sglang/srt/layers/quantization/unquant.py:609]: why we need this `drop_pad_mode`?

Comment (Vladimir221) [reply python/sglang/srt/layers/quantization/unquant.py:609]: It seems `torch_npu.npu_moe_init_routing_v2` returns `expanded_row_idx` in another format ([torch_npu-npu_moe_finalize_routing.doc](https://www.hiascend.com/document/detail/zh/Pytorch/730/apiref/torchnpuCustomsapi/docs/context/torch_npu-npu_moe_finalize_routing.md)), so without `drop_pad_mode = 2` `torch_npu-npu_moe_finalize_routing` leads to incorrect result

File: python/sglang/srt/layers/quantization/unquant.py (hunk 1 above)
Comment (Vladimir221) [reply python/sglang/srt/layers/quantization/unquant.py]: Will change after #15904 is merged

File: python/sglang/srt/layers/quantization/unquant.py [hunk 3]
```diff
@@ -308,6 +308,9 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
                 weight.data = npu_format_cast(
                     weight.data,
                 )
+            if layer.w13_weight.shape[-1] == layer.hidden_size:
```
Comment (OrangeRedeng) [inline python/sglang/srt/layers/quantization/unquant.py]: Please remove this strings

Comment (OrangeRedeng) [reply python/sglang/srt/layers/quantization/unquant.py]: they are no longer needed after https://github.com/sgl-project/sglang/pull/15904 merge

Comment (Vladimir221) [reply python/sglang/srt/layers/quantization/unquant.py]: Done

**Review by iforgetmyname — APPROVED** [review]

File: python/sglang/srt/layers/quantization/unquant.py [hunk 4]
```diff
@@ -543,35 +542,33 @@ def forward_npu(
...
+            )
         )
-
         expert_tokens = expert_tokens.to(torch.int64)
         w13_bias = [layer.w13_weight_bias] if self.with_bias else None
         w2_bias = [layer.w2_weight_bias] if self.with_bias else None
 
         # gmm1: gate_up_proj
-        hidden_states = torch_npu.npu_grouped_matmul(
+        hidden_states = torch.ops.npu.npu_grouped_matmul(
```
Comment (silencejade) [inline python/sglang/srt/layers/quantization/unquant.py:566]: What are the differences between torch_npu and torch.ops.npu?

Comment (Vladimir221) [reply python/sglang/srt/layers/quantization/unquant.py:566]: I suppose they should be equal interfaces but torch_npu is deprecated I think
image
